### PR TITLE
[qa] Add a check for empty line #29

### LINKED ISSUE #29: [qa] Add a check for empty line at the end of a file
We should add another automated test to check if all the files have an empty line at the end of the file.
Many GCI students are making this mistake and an automated test will help them in my opinion! :smile: 

PS: **binary files must be ignored**, the check should be performed only on text files.
If it can be performed with just a series of shell commands, we can use that and start adding this check to the `.travis.yml` build of this repo

### PR DESCRIPTION
**Pull request by unsuitable001**
This update adds a check for empty line is present at the end of all files in the repository
GCI Task - https://codein.withgoogle.com/dashboard/task-instances/4670648862375936/

Fixes #29

### CODE REVIEW

File: endline-script.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+filelist = subprocess.check_output('find . -type f ' + excludepath, cwd=cwdn, shell=True)
+filelist = filelist.decode('utf-8')
+pathvar = ''
+for c in filelist:
+    if(c != '\n'):
+        pathvar = pathvar + c
+    else:
+        check = subprocess.check_output('tail -c 1 ' + pathvar, cwd=cwdn, shell=True)
+        if(check.decode('utf-8') != '\n'):
+            print(pathvar + " Dont have a blank line at the end")
```
Comment (atb00ker) [inline endline-script.py]: `pathvar + " doesn't have a blank line at the end of the file"`

Comment (unsuitable001) [reply endline-script.py]: Ok thanks :) sorry for my bad English

**Comment by ppabcd** [comment]
@unsuitable001 don't forget to create unit testing .-.

**Comment by unsuitable001** [comment]
Ok. I'll do it. :) Other things are fine nah?

File: openwisp_utils/qa.py [hunk 2]
```diff
@@ -205,3 +206,47 @@ def _find_issue_mentions(message):
...
+    a blank endline after it
+    """
+    args = _parse_endline_check_args()
+    excludepath = '! -path "../.*" ! -path "*.pyc" ! -path "*__.py" ! -path "*.svg" ! -path "*.png" '
+    excludepath = excludepath + '! -path "*/*.egg-info/*" ! -path "../.*/*" ! -path "../*/.*" '
+    excludepath = excludepath + '! -path "*/test*/*"'
+    if args.skipscript is None:
+        pass
+    else:
+        excludepath = excludepath + str(args.skipscript)
```
[the remark this answers is not in the document]
Comment (unsuitable001) [reply openwisp_utils/qa.py]: Can you suggest anything about it?

[the remark this answers is not in the document]
Comment (unsuitable001) [reply openwisp_utils/qa.py]: Can you suggest anything about it?

File: openwisp_utils/qa.py [hunk 3]
```diff
@@ -205,3 +206,47 @@ def _find_issue_mentions(message):
...
+    # QA Check
+    cwdn = os.getcwd()
+    filelist = subprocess.check_output('find .. -type f ' + excludepath, cwd=cwdn, shell=True)
+    filelist = filelist.decode('utf-8')
+    pathvar = ''
+    for c in filelist:
```
[the remark this answers is not in the document]
Comment (unsuitable001) [reply openwisp_utils/qa.py]: Ok

**Comment by unsuitable001** [comment]
I'll do that tomorrow.

File: .coveragerc [hunk 4]
```diff
@@ -5,3 +5,4 @@ omit =
     /*/__init__.py
     /setup.py
     /*/migrations/*
+
```
[the remark this answers is not in the document]
Comment (unsuitable001) [reply .coveragerc]: actually, there was an empty new line after those spaces. So, the script was not detecting it.
I'm removing the space :)

**Comment by unsuitable001** [comment]
`about qa.py`
In case you are wondering, why I did
` i = 0
    while(i < listlength - 1):`
instead of 
`for pathvar in filelist`

the last element of `filelist` is just empty. when `tail` gets that element, it just freezes. So, the build fails. That's why I have used `while loop` & skipped the `last element of filelist`

**Comment by cking100** [comment]
@unsuitable001 I think this is not a good way of doing this a better way would be to use list slicing something like this
```
for file in filelist[:-1]:
```
Or why not just pop the empty string from the list?

**Comment by unsuitable001** [comment]
Yeah, that's also a good idea. But, why running while loop isn't good?

**Comment by unsuitable001** [comment]
@atb00ker what do you say? If you want me to pop it from list or use the for loop as above, then, tell me here. No need to do that in gci system. I'll do it. :)

**Comment by unsuitable001** [comment]
@atb00ker changes done! Now, you can review.

**Comment by unsuitable001** [comment]
> @unsuitable001 I think this is not a good way of doing this a better way would be to use list slicing something like this
> 
> ```
> for file in filelist[:-1]:
> ```
> 
> Or why not just pop the empty string from the list?

thanks for your concern!

**Comment by cking100** [comment]
@unsuitable001 refer the tests written for `check_migrations_name` and `check_commit_message`

**Comment by unsuitable001** [comment]
> @unsuitable001 refer the tests written for `check_migrations_name` and `check_commit_message`

I have 0 knowledge about these coverage thing. Can you tell me on which files I should observe? & where to write the actual test code?

**Comment by cking100** [comment]
First of all check the coverage reported by the coveralls and go to the file which is resulting in a decrease you will see the red lines indicated which are not tested. There is no point if i would tell you the exactly location of where to put your code. Just have a look on tests and see where is `check_commit_message` test is written, you have to write test in that file itself @unsuitable001

**Comment by unsuitable001** [comment]
> First of all check the coverage reported by the coveralls and go to the file which is resulting in a decrease you will see the red lines indicated which are not tested. There is no point if i would tell you the exactly location of where to put your code. Just have a look on tests and see where is `check_commit_message` test is written, you have to write test in that file itself @unsuitable001

I saw the log. It was the previous version of the file :worried: . The new file isn't there yet. :worried:

**Comment by cking100** [comment]
> [Coverage Status](https://coveralls.io/builds/20487753)
> 
> Coverage decreased (-10.7%) to 87.081% when pulling **[e520563](https://github.com/openwisp/openwisp-utils/commit/e520563fbd5734beffc8f9a504504a30a1805e92) on unsuitable001:master** into **[921cc11](https://github.com/openwisp/openwisp-utils/commit/921cc11a544943ac7f5c088da631987161c5e9c6) on openwisp:master**.

Just click on the coverage icon

**Comment by unsuitable001** [comment]
> > [Coverage Status](https://coveralls.io/builds/20487753)
> > Coverage decreased (-10.7%) to 87.081% when pulling **[e520563](https://github.com/openwisp/openwisp-utils/commit/e520563fbd5734beffc8f9a504504a30a1805e92) on unsuitable001:master** into **[921cc11](https://github.com/openwisp/openwisp-utils/commit/921cc11a544943ac7f5c088da631987161c5e9c6) on openwisp:master**.
> 
> Just click on the coverage icon

I know. I clicked that. The file that was showing there was the previous version of my file. Not the current one. You can also compare coveralls' file and github's changed files tab

**Comment by unsuitable001** [comment]
This coveralls thing is looking too confusing to me. same lines used above are green. But, red in the later lines.

**Comment by cking100** [comment]
> > > [Coverage Status](https://coveralls.io/builds/20487753)
> > > Coverage decreased (-10.7%) to 87.081% when pulling **[e520563](https://github.com/openwisp/openwisp-utils/commit/e520563fbd5734beffc8f9a504504a30a1805e92) on unsuitable001:master** into **[921cc11](https://github.com/openwisp/openwisp-utils/commit/921cc11a544943ac7f5c088da631987161c5e9c6) on openwisp:master**.
> > 
> > 
> > Just click on the coverage icon
> 
> I know. I clicked that. The file that was showing there was the previous version of my file. Not the current one. You can also compare coveralls' file and github's changed files tab.
> 
> I think, this report is for any previous commits I had made.

For that you will have to build in your local machine and you will be able to run coverage from command line

**Comment by cking100** [comment]
@unsuitable001 it is better to build the project and check coverage from command line so do that and if you need any help in that PM me on gitter

File: openwisp_utils/qa.py [hunk 5]
```diff
@@ -205,3 +206,42 @@ def _find_issue_mentions(message):
...
+    excludepath = excludepath + '! -path "*/*.egg-info/*" ! -path "../.*/*" ! -path "../*/.*" '
+    excludepath = excludepath + '! -path "*/test*/*" '
+    if args.skipscript is not None:
+        excludepath = excludepath + str(args.skipscript)
+    # QA Check
+    cwdn = os.getcwd()
+    filelist = subprocess.check_output('find .. -type f ' + excludepath, cwd=cwdn, shell=True)
+    filelist = filelist.decode('utf-8')
+    filelist = filelist.split('\n')
+    for pathvar in filelist[: - 1]:
```
Comment (atb00ker) [inline openwisp_utils/qa.py]: `:-1` (remove spaces)

File: openwisp_utils/qa.py [hunk 6]
```diff
@@ -205,3 +206,42 @@ def _find_issue_mentions(message):
...
+        check = subprocess.check_output('tail -c 1 ' + pathvar, cwd=cwdn, shell=True)
+        try:
+            if(check.decode('utf-8') != '\n'):
+                print(pathvar + " doesn't have a blank line at the end")
+                sys.exit(1)
+        except UnicodeDecodeError:
+            print(pathvar + " is a binary file")
```
Comment (atb00ker) [inline openwisp_utils/qa.py]: I don't think printing that it's a binary file will help. Any particular reason for this?

File: tests/test_project/tests/test_qa.py [hunk 7]
```diff
@@ -184,5 +185,78 @@ def test_qa_call_check_commit_message_failure(self):
...
+        options = [
+            [
+                'checkendline',
+                '--skipscript',
+                "! -path file.jpg"
+            ],
+            [
+                'checkendline',
+                '--skipscript',
+                "!-path 'file.tiff'"
```
Comment (atb00ker) [inline tests/test_project/tests/test_qa.py]: `!-path` -> `! -path`

File: openwisp_utils/qa.py (hunk 6 above)
Comment (unsuitable001) [reply openwisp_utils/qa.py]: no. just for debugging purposes. I'll delete it

**Comment by unsuitable001** [comment]
@nemesisdesign i'll talk to you about this after gci :)

**Comment by unsuitable001** [comment]
@nemesisdesign 
> he exclude mechanism doesn't take in consideration the possibility of binary files at all, it 

I have excluded some file formats. That's true. But, at the end of the code, there's a exception check `UnicodeDecodeError` . When, script checks any binary file (except excluded) will raise a `UnicodeDecodeError` and skip it. But, you have to add skip for editable files like `.svg`

> this code is basically doing a shell check in python, so there's no rea...


I am more comfortable in python and I thought, python is allowed. And, more more reason to take python. How can I skip binary files?
